dask/array/tests/test_gufunc.py

```diff
@@ -0,0 +1,262 @@
+    def foo(x, bar):
+        assert bar == 2
+        return x
+    ret = apply_gufunc(foo, "()->()", 1., output_dtypes="f", bar=2)
+    assert_eq(ret, np.array(1., dtype="f"))
+
+
+@pytest.mark.xfail(reason="Currently np.einsum doesn't seem to broadcast correctly for this case")
+def test_apply_gufunc_02():
+    def outer_product(x, y):
+        return np.einsum("...i,...j->...ij", x, y)
+    a = da.random.normal(size=(   20, 30), chunks=(5, 30))
+    b = da.random.normal(size=(10, 1, 40), chunks=(10, 1, 40))
+    c = apply_gufunc(outer_product, "(i),(j)->(i,j)", a, b, output_dtypes=a.dtype)
+    assert c.compute().shape == (10, 20, 30, 40)
+
+
+def test_apply_gufunc_scalar_output():
+    def foo():
+        return 1
+    x = apply_gufunc(foo, "->()", output_dtypes=int)
+    assert x.compute() == 1
+
+
+def test_apply_gufunc_elemwise_01():
+    def add(x, y):
+        return x + y
+    a = da.from_array(np.array([1, 2, 3]), chunks=2, name='a')
+    b = da.from_array(np.array([1, 2, 3]), chunks=2, name='b')
+    z = apply_gufunc(add, "(),()->()", a, b, output_dtypes=a.dtype)
+    assert_eq(z, np.array([2, 4, 6]))
+
+
+def test_apply_gufunc_elemwise_01b():
+    def add(x, y):
+        return x + y
+    a = da.from_array(np.array([1, 2, 3]), chunks=2, name='a')
+    b = da.from_array(np.array([1, 2, 3]), chunks=1, name='b')
+    with assert_raises(ValueError):
+        apply_gufunc(add, "(),()->()", a, b, output_dtypes=a.dtype)
+
+
+def test_apply_gufunc_elemwise_02():
+    def addmul(x, y):
+        assert x.shape in ((2,), (1,))
+        return x + y, x * y
+    a = da.from_array(np.array([1, 2, 3]), chunks=2, name='a')
+    b = da.from_array(np.array([1, 2, 3]), chunks=2, name='b')
+    z1, z2 = apply_gufunc(addmul, "(),()->(),()", a, b, output_dtypes=2 * (a.dtype,))
+    assert_eq(z1, np.array([2, 4, 6]))
+    assert_eq(z2, np.array([1, 4, 9]))
+
+
+def test_gufunc_vector_output():
+    def foo():
+        return np.array([1, 2, 3], dtype=int)
+    x = apply_gufunc(foo, "->(i_0)", output_dtypes=int, output_sizes={"i_0": 3})
+    assert x.chunks == ((3,),)
+    assert_eq(x, np.array([1, 2, 3]))
+
+
+def test_apply_gufunc_elemwise_loop():
+    def foo(x):
+        assert x.shape in ((2,), (1,))
+        return 2 * x
+    a = da.from_array(np.array([1, 2, 3]), chunks=2, name='a')
+    z = apply_gufunc(foo, "()->()", a, output_dtypes=int)
+    assert z.chunks == ((2, 1),)
+    assert_eq(z, np.array([2, 4, 6]))
+
+
+def test_apply_gufunc_elemwise_core():
+    def foo(x):
+        assert x.shape == (3,)
+        return 2 * x
+    a = da.from_array(np.array([1, 2, 3]), chunks=3, name='a')
+    z = apply_gufunc(foo, "(i)->(i)", a, output_dtypes=int)
+    assert z.chunks == ((3,),)
+    assert_eq(z, np.array([2, 4, 6]))
+
+
+# TODO: In case single tuple output will get enabled:
+# def test_apply_gufunc_one_scalar_output():
+#     def foo():
+#         return 1,
+#     x, = apply_gufunc(foo, "->(),", output_dtypes=(int,))
```

Comment:
OK, now I understand why you special cased one output. Remind me, did you open an upstream issue in NumPy to discuss this?

My guess is that even though this is arguably poor API (special casing one vs. many return values) there will be little appetite for changing/extending it in NumPy itself.

Reply:
Yes, the upstream issue is numpy/numpy#10527

I slowly get to understand that I might need to convince many people for this to change upstream and that the order should be to implement against current API and change later if upstream changes. 

Of course all your other code comments vanish, if I'd roll back to the current signature API. 

As mentioned before I'd like to be transparent about the current implementatio, which means also (focus on `output_dtypes`):
```python
def foo(x):
    return x
a = np.array([1, 2, 3])
apply_gufunc(foo, "()->()", a, output_dtypes=int)  # Not: output_dtypes=[int]
```

What do you recommend/favor? Keep or roll back to current signature API?